qiskit/extensions/unitary.py

```diff
@@ -106,6 +111,55 @@ def _define(self):
             raise NotImplementedError("Not able to generate a subcircuit for "
                                       "a {}-qubit unitary".format(self.num_qubits))
 
+    def qasm(self):
+        """ The qasm for a custom unitary gate
+        This is achieved by adding a custom gate that corresponds to the definition
+        of this gate. It gives the gate a random name if one hasn't been given to it.
+        """
+        # if this is true then we have written the gate definition already
+        # so we only need to write the name
+        if self.qasm_def_written:
+            return self._qasmif(self.qasm_name)
+
+        # we have worked out the definition before, but haven't written it yet
+        # so we need to write definition + name
+        if self.qasm_definiton:
+            self.qasm_def_written = True
```

Comment:
Can this be `self._qasm_def_written`?

Reply:
I added 3 new fields, and they all should be private which I have done in 4d3aff9